Clustering/resource management

cryogenian opened the pull request:
[ch10188] [ch10325] 

Sorry for formatting diffs, but I wrote these files from scratch and then replace old versions with new ones not modified existing stuff. 

Basically this does the following 

+ Introduces `ResourceManager`. This is something like generalization of `DatasourceManager` and `DestinationManager` but can't do anything fancy, just `manage`, `shutdown` and `get` by id. 
+ Introduces `CachedGetter`, small wrapper around memoized `I => F[Option[A]]` that allows to handle indexed store content changes be handled in outer world. 
+ Introduces `WhateverModules` as an abstraction used to build resources of datasources and destinations. These modules take care about all those `CreateError` and so on. 
+ Removes `DatasourceManager` and `DestinationManager`, `DefaultDatasources` and `DefaultDestinations` 
+ Reintroduces `DefaultDatasources` and `DefaultDestinations` using resource manager cache getter and `WhateverModules` (note, that destinations part is kinda strange a bit, as it has errors in parameters but it's not set anywhere, I preserved this, but not sure if this is intended) 
+ Makes `DatasourcesSpec` working with `Resource[F, Datasources[...]]` instead of pure datasources 
+ Adds tests for store being more important than cache.

cryogenian commented:
@wemrysi @jsantos17 Ping 🙂

wemrysi commented:
@cryogenian Sorry for the delay, I'll finish reviewing today.

jsantos17 commented on impl/src/main/scala/quasar/impl/CachedGetter.scala:
```diff
@@ -0,0 +1,66 @@
+  def apply[F[_]: Sync, I, A: Eq](getter: I => F[Option[A]]): F[CachedGetter[F, I, A]] =
+    Ref.of[F, Map[I, A]](Map.empty) map { (cache: Ref[F, Map[I, A]]) =>
+      new CachedGetter[F, I, A] {
+        def apply(i: I): F[Signal[A]] = for {
```
Any reason why the implementation is here instead of being in its own class?

cryogenian replied:
There is no but the class would be very small and its constructor will be private. Why would we need it?

cryogenian replied:
Discussed in discord, @jsantos17 is fine with anonymous classes here

jsantos17 commented on impl/src/main/scala/quasar/impl/ResourceManager.scala:
```diff
@@ -0,0 +1,60 @@
+import cats.effect.{Resource, Sync}
+import cats.effect.concurrent.Ref
+import cats.effect.syntax.bracket._
+import cats.syntax.applicative._
```
Not requesting any changes, but you could just import `cats.implicits._` if you want.

jsantos17 commented on impl/src/main/scala/quasar/impl/datasources/DefaultDatasources.scala:
```diff
@@ -16,204 +16,232 @@
-      case None =>
-        DatasourceError.datasourceNotFound[I, E](datasourceId)
-          .left[A].point[F]
+  private def getMDS[E >: ExistentialError[I] <: DatasourceError[I, C]](i: I): EitherT[F, E, MDS] = {
```
What is `MDS`?

cryogenian replied:
A type synonim for `ManagedDatasource[...]` used in this module, do you think it would be better if the aliasing happens a bit closer to this method?

jsantos17 commented on impl/src/main/scala/quasar/impl/ResourceManager.scala:
```diff
@@ -0,0 +1,60 @@
+import cats.instances.option._
+
+trait ResourceManager[F[_], I, A] {
+  def manage(i: I, allocated: (A, F[Unit])): F[Unit]
```
Can we get some commnets on what each of these do? I'm not sure I understand what `manage` does?

cryogenian replied:
:+1:

cryogenian commented on impl/src/main/scala/quasar/impl/ResourceManager.scala:
```diff
@@ -0,0 +1,60 @@
+
+object ResourceManager {
+  def apply[F[_]: Sync, I, A]: Resource[F, ResourceManager[F, I, A]] = {
+    val fPair = Ref.of[F, Map[I, (A, F[Unit])]](Map.empty) map { ref =>
```
:+1: for thread safety. Why do we need it to be concrete class? How is it better?

jsantos17 replied:
I mean in the implementation. It's fine to keep the abstract type in the signature

cryogenian replied:
Discussed in discord, @jsantos17 is fine with anonymous classes here

cryogenian commented on impl/src/test/scala/quasar/impl/CachedGetterSpec.scala:
```diff
@@ -0,0 +1,61 @@
+
+class CachedGetterSpec(implicit ec: ExecutionContext) extends EffectfulQSpec[IO] {
+  import CachedGetter.Signal._
+  "CachedGetter works" >>* {
```
:+1: I didn't come up with nice message, it tests that `CachedGetter` behaves as expected.

cryogenian commented on impl/src/test/scala/quasar/impl/datasources/DatasourceModulesSpec.scala:
```diff
@@ -0,0 +1,270 @@
+      }
+    }
+  }
+  "sanitizing refs" >>* {
```
That `DatasourceModules` sanitizes refs 🙂

wemrysi commented on api/src/test/scala/quasar/api/datasource/DatasourcesSpec.scala:
```diff
@@ -59,43 +60,48 @@ abstract class DatasourcesSpec[
+  def mutationExamples(f: (Datasources[F, G, I, C, S], DatasourceRef[C]) => F[Err \/ I]) = {
     "lookup on success" >>* {
       for {
+        (dses, finalize) <- datasources.allocated
```
How about wrapping the `for` blocks in `datasources.run(...)`?

cryogenian commented:
@jsantos17 @wemrysi I think it's ready for another look

+ Made `create` returning `EitherT[Resource[F, ?], CreateError[C], MDS]` 
+ Changed `CachedGetter` test message
+ Moved `MDS` alias closer to `getMDS` method
+ Use `datasource.use` in `DatasourcesSpecs` 
+ Updated to master

wemrysi commented on impl/src/main/scala/quasar/impl/ResourceManager.scala:
```diff
@@ -39,24 +39,36 @@ trait ResourceManager[F[_], I, A] {
+  def apply[F[_]: Concurrent, I, A]: Resource[F, ResourceManager[F, I, A]] = {
+    val fPair = for {
+      ref <- Ref.of[F, Map[I, (A, F[Unit])]](Map.empty)
+      semaphore <- Semaphore[F](1)
```
You might also consider replacing `Ref` + `Semaphore` with an `MVar`.

wemrysi replied:
I guess that'd be slightly different semantics as reads would block on any in-progress mutation.

cryogenian replied:
Right! I considered this and didn't switch to `MVar` because of read

cryogenian replied:
TBH, I'm not sure if this is correct behaviour or not 🙂 It looks like the main problem to handle is conflicting updates, not reads, I can switch to `MVar` if you think it's better

wemrysi replied:
I think the current impl is probably preferable to a single `MVar`, we'd need the equivalent of an `MVar` per-index to minimize blocking and prevent reading a value in the process of being modified.

alissapajer commented:
Chatted with @cryogenian and I'm going to take charge of merging this PR, and all the upstream ones, after my series of breaking `RateLimiter` PRs lands. 🛑 is up for that reason.
